### PR TITLE
Make bwbcoverbot use cover info from ol dumps

### PR DESCRIPTION
**Pull request by scottbarnes**
Closes no issue.

## Description:
Added some tests to BWBCoverBot, and a little bit of documentation to the README. The behavior should remain the same. This is a work in progress.

### CODE REVIEW

File: BWBCoverBot/README.md [hunk 1]
```diff
@@ -1 +1,21 @@
...
+A script to update covers on Open Library using a zip of covers from BWB.
+
+# Setup
+Export `OL_ACCESS_KEY` and `OL_SECRET_KEY` (s3) used by the [BWBImportBot](https://openlibrary.org/people/bwbimportbot) account to import.
+
+# Use
+Run `python main.py bwbcoverfile.zip`. E.g. `python main.py Apr2021_17_lc_13.zip`.
+
+# How it works
+Each zip file from BWB contains hundreds of `.jpg` files, each prefixed with an ISBN 13 (e.g. `9781483457550.jpg`).
```
Comment (cclauss) [inline BWBCoverBot/README.md:11]:
```suggestion
Each zip file from BWB contains hundreds of `.jpg` files.  The root of each filename is an ISBN 13 (e.g. `9781483457550.jpg`).
```

File: BWBCoverBot/README.md [hunk 2]
```diff
@@ -1 +1,21 @@
...
+
+BWBCoverBot goes through the zip and for each `.jpg` it checks `./bwb-cover-bot.sqlite` to see if the associated ISBN 13 from the `.jpg`'s filename is in the DB, and if `cover_exists` is 1.
+
+If the cover is already in the DB, BWBCoverBot moves on, but if it's not, it will query the Open Library API, check if there is a cover, add the cover if necessary, and update `cover_exists`.
```
Comment (cclauss) [inline BWBCoverBot/README.md:15]:
```suggestion
BWBCoverBot goes through the zip and for each `.jpg` it checks `./bwb-cover-bot.sqlite` to see if the associated ISBN 13 from the filename is in the DB, and if `cover_exists` is not 1.

If the cover is not already in the DB, BWBCoverBot will query the Open Library API, check if there is a cover, add the cover if necessary, and set `cover_exists` to 1.
```

File: BWBCoverBot/conftest.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import pytest
+from requests_mock.mocker import Mocker as RequestsMock
+from sqlmodel import SQLModel, create_engine, Session
+from olclient import OpenLibrary, config
+from main import EditionCoverData
+
+
+@pytest.fixture(scope="session")
+def get_db():
```
Comment (cclauss) [inline BWBCoverBot/conftest.py:9]: Add a type hint for the yielded data?

File: BWBCoverBot/conftest.py [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    db_session.bulk_save_objects(
+        [EditionCoverData(isbn_13="123", cover_exists=True)]
+    )
+    db_session.commit()
+
+    yield db_session
+
+
+@pytest.fixture()
+def get_ol(requests_mock: RequestsMock):
```
Comment (cclauss) [inline BWBCoverBot/conftest.py:24]: Add a type hint for the yielded data?

File: BWBCoverBot/main.py [hunk 5]
```diff
@@ -15,11 +15,12 @@
     handlers=[logging.FileHandler("bwb-cover-bot-debug.log"), logging.StreamHandler()],
 )
 
-ol = OpenLibrary(
-    credentials=config.Credentials(
-        access=os.environ["OL_ACCESS_KEY"], secret=os.environ["OL_SECRET_KEY"]
+def get_ol():
```
Comment (cclauss) [inline BWBCoverBot/main.py:18]: Add a type hint for the yielded data?

File: BWBCoverBot/main.py [hunk 6]
```diff
@@ -57,17 +58,18 @@ def is_cover_already_stored(isbn_13: str) -> bool:
     )
 
 
-def verify_and_update_cover(isbn_13: str, archive_contents: ZipFile) -> None:
+def verify_and_update_cover(isbn_13: str, archive_contents: ZipFile, ol: OpenLibrary) -> None:
```
Comment (cclauss) [inline BWBCoverBot/main.py:61]: Would it be helpful if this function returned a `bool` to let the caller know if the cover was verified and updated?

File: BWBCoverBot/main.py [hunk 7]
```diff
@@ -89,7 +93,7 @@ def verify_and_update_cover(isbn_13: str, archive_contents: ZipFile) -> None:
     return
 
 
-def parser_for_zip_with_isbns(cover_zip_path: str) -> None:
+def parser_for_zip_with_isbns(cover_zip_path: str, ol: OpenLibrary) -> None:
```
Comment (cclauss) [inline BWBCoverBot/main.py:96]: Would it be helpful if this function returned an `int` to let the caller know how many covers were verified and updated?

File: BWBCoverBot/tests/test_main.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+        requests_mock: RequestsMock, get_ol: OpenLibrary
+) -> bool | None:
+    """Check that correct data is POSTed to the backend."""
+    m = requests_mock
+    olid = "OL1234M"
+    mime_type = "image/jpeg"
+    file_name = "1234567890123.jpg"
+    cover_fp = open(f"./tests/{file_name}", "rb")
+    cover_data = cover_fp.read()
+    cover_fp.close()
```
Comment (cclauss) [inline BWBCoverBot/tests/test_main.py:28]: Improvement.  Use `with open() as` to ensure the file handle is closed even if an exception is raised.

Bigger improvement.  Use [pathlib's read_bytes()](https://docs.python.org/3/library/pathlib.html#pathlib.Path.read_bytes).

**Review by cclauss — APPROVED** [review]

File: BWBCoverBot/README.md (hunk 1 above)
Comment (scottbarnes) [reply BWBCoverBot/README.md]: I made this change.

File: BWBCoverBot/README.md (hunk 2 above)
Comment (scottbarnes) [reply BWBCoverBot/README.md]: I made this change.

File: BWBCoverBot/conftest.py (hunk 3 above)
Comment (scottbarnes) [reply BWBCoverBot/conftest.py]: I added a type hint here.

File: BWBCoverBot/conftest.py (hunk 4 above)
Comment (scottbarnes) [reply BWBCoverBot/conftest.py]: I added a type hint here.

File: BWBCoverBot/main.py (hunk 6 above)
Comment (scottbarnes) [reply BWBCoverBot/main.py]: I added a bool here.

File: BWBCoverBot/main.py (hunk 7 above)
Comment (scottbarnes) [reply BWBCoverBot/main.py]: I did this, I hope in line with what you were thinking. :)

File: BWBCoverBot/tests/test_main.py (hunk 8 above)
Comment (scottbarnes) [reply BWBCoverBot/tests/test_main.py]: Good call! I went with pathlib.

**Comment by scottbarnes** [comment]
@cclauss, I hope these changes fit with what you were thinking. I tried this out in my local environment, but I can't get the script to completely run locally, either as it was before, or now, so it probably needs testing against a real BWB cover zip.

File: BWBCoverBot/main.py (hunk 5 above)
Comment (scottbarnes) [reply BWBCoverBot/main.py]: I added this.

**Comment by cclauss** [comment]
Please add `sqlmodel` to https://github.com/internetarchive/openlibrary-bots/blob/master/requirements.txt to placate the pytest.

**Comment by scottbarnes** [comment]
I think it's happier now. :)

**Comment by cclauss** [comment]
Please rebase so that the tests pass.

**Comment by scottbarnes** [comment]
@cclauss, I rebased, but the tests here don't seem to run.

**Comment by cclauss** [comment]
ModuleNotFoundError: No module named 'requests_mock'

**Comment by scottbarnes** [comment]
@cclauss, I moved the requests-mock and another development dependency from `BWBCoverBot/requirements-dev.txt` to `BWBCoverBot/requirements.txt`.

**Comment by cclauss** [comment]
Please add your requirements file to:
https://github.com/internetarchive/openlibrary-bots/blob/master/.github/workflows/lint_python.yml#L20
And then you can remove `sqlmodel` from the main `requirements.txt`

We can work on pytest test discovery after the requirements are loading properly.

**Comment by cclauss** [comment]
`requirements.txt` should be sorted in alphabetical order by dependency name where possible.  Doing so makes it easy to spot missing items and nearly impossible to add a duplicate item. When lists are short, this will seem like overkill but as they get longer sorting becomes increasingly valuable.

File: BWBCoverBot/tests/test_main.py [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+
+
+@pytest.mark.parametrize("tc,exp", [(tc, tc.exp) for tc in test_cases])
+def test_verify_and_update_cover(
+    requests_mock: RequestsMock, get_db: Session, get_ol: OpenLibrary, tc, exp, caplog
+) -> None:
+    """See the "text" field in test_cases for an explanation of each test case."""
+
+    m = requests_mock
+    archive_contents = ZipFile("./tests/test_isbns.zip", "r")
```
Comment (cclauss) [inline BWBCoverBot/tests/test_main.py:146]: Use `pathlib` to join `__file__` with `test_isbns.zip` to define the path to this file.

Comment (cclauss) [reply BWBCoverBot/tests/test_main.py:146]: [`__file__`](https://docs.python.org/3/reference/import.html#__file__)

Comment (cclauss) [reply BWBCoverBot/tests/test_main.py:146]: test_isbns_zip = pathlib.Path(__file__).parent.resolve() / "test_isbns.zip"

**Comment by scottbarnes** [comment]
@cclauss, sorry that took so long. There was an issue with `caplog` and the log level being overwritten somehow (I think), such that the tests would pass with `pytest BWBCoverBot`, but not `pytest`, without the path.

I 'fixed' it by setting `caplog.set_level(logging.INFO)` in the tests.

File: BWBCoverBot/tests/test_main.py (hunk 9 above)
Comment (scottbarnes) [reply BWBCoverBot/tests/test_main.py]: Ahhh, good call. Thanks!

**Pull request merged**